<a href="https://colab.research.google.com/github/vnzhere/Python-/blob/main/Housing_marketing_intelligence.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Housing Market Intelligence & Price Prediction System**

 A Complete Data Analysis & Machine Learning Project


**Goal:** Understand what drives housing prices and build a predictive model  

**Tools:** Python · Pandas · NumPy · Matplotlib · Scikit-learn  

**Author:** Vanzhere
**Date:** May 2026

**STEP 1: DATA LOADING**

1. Import libraries

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%3f' % x)

print("Libraries imported!")

Libraries imported!


2. Load Dataset

In [7]:
import pandas as pd

url = "https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv"

df = pd.read_csv(url)

print("Dataset successfully loaded!")

Dataset successfully loaded!


**STEP 2: DATA UNDERSTANDING**

Why do we explore before we analyze data?


* Data is unknown which on jumping directly to analysis means driving without a map.
* These commands helps us in understanding, giving structure and plan our data cleaning strategy




In [8]:
df.head()
#gives column names, data types and sample values

,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,b,lstat,medv
0,0.006320,18.000000,2.310000,0,0.538000,6.575000,65.200000,4.090000,1,296,15.300000,396.900000,4.980000,24.000000
1,0.027310,0.000000,7.070000,0,0.469000,6.421000,78.900000,4.967100,2,242,17.800000,396.900000,9.140000,21.600000
2,0.027290,0.000000,7.070000,0,0.469000,7.185000,61.100000,4.967100,2,242,17.800000,392.830000,4.030000,34.700000
3,0.032370,0.000000,2.180000,0,0.458000,6.998000,45.800000,6.062200,3,222,18.700000,394.630000,2.940000,33.400000
4,0.069050,0.000000,2.180000,0,0.458000,7.147000,54.200000,6.062200,3,222,18.700000,396.900000,5.330000,36.200000


In [9]:
df.tail()
#tells us about data changes in the end (common in time ordered datasets)

,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,b,lstat,medv
501,0.062630,0.000000,11.930000,0,0.573000,6.593000,69.100000,2.478600,1,273,21.000000,391.990000,9.670000,22.400000
502,0.045270,0.000000,11.930000,0,0.573000,6.120000,76.700000,2.287500,1,273,21.000000,396.900000,9.080000,20.600000
503,0.060760,0.000000,11.930000,0,0.573000,6.976000,91.000000,2.167500,1,273,21.000000,396.900000,5.640000,23.900000
504,0.109590,0.000000,11.930000,0,0.573000,6.794000,89.300000,2.388900,1,273,21.000000,393.450000,6.480000,22.000000
505,0.047410,0.000000,11.930000,0,0.573000,6.030000,80.800000,2.505000,1,273,21.000000,396.900000,7.880000,11.900000


In [10]:
#gives dimensions(r,c)
print(f"Shape: {df.shape}")

# means total no. of rows
# each row represents one house/ data obsv.
print(f"→ {df.shape[0]} rows (houses/observations)")

# total no. of columns
# represents attributes/feature of a house
print(f"→ {df.shape[1]} columns (features + 1 target)")

#returns colm names in form of a list
print(f"\nColumn names:\n{list(df.columns)}")

# "medv"-- is the value we want to predict/analyze (median home values in 1000s)
print(f"\nTarget variable: 'medv' (Median Home Value in $1000s)")

# remaining colms are feature variables, basically are inputs used to analyze/predict target variables
print(f"Feature variables: all other 13 columns")

Shape: (506, 14)
→ 506 rows (houses/observations)
→ 14 columns (features + 1 target)

Column names:
['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'b', 'lstat', 'medv']

Target variable: 'medv' (Median Home Value in $1000s)
Feature variables: all other 13 columns


In [11]:
# tells how many non-null values each column has
# datatype of each colm
# total memory usage

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  b        506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


In [12]:
# gives 8 statistical parameters for every numeric column

# if mean~median ---- data is normally distributed
# if very diff.---- data is skewed, outliers exist
df.describe()


,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,b,lstat,medv
count,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000,506.000000
mean,3.613524,11.363636,11.136779,0.069170,0.554695,6.284634,68.574901,3.795043,9.549407,408.237154,18.455534,356.674032,12.653063,22.532806
std,8.601545,23.322453,6.860353,0.253994,0.115878,0.702617,28.148861,2.105710,8.707259,168.537116,2.164946,91.294864,7.141062,9.197104
min,0.006320,0.000000,0.460000,0.000000,0.385000,3.561000,2.900000,1.129600,1.000000,187.000000,12.600000,0.320000,1.730000,5.000000
25%,0.082045,0.000000,5.190000,0.000000,0.449000,5.885500,45.025000,2.100175,4.000000,279.000000,17.400000,375.377500,6.950000,17.025000
50%,0.256510,0.000000,9.690000,0.000000,0.538000,6.208500,77.500000,3.207450,5.000000,330.000000,19.050000,391.440000,11.360000,21.200000
75%,3.677083,12.500000,18.100000,0.000000,0.624000,6.623500,94.075000,5.188425,24.000000,666.000000,20.200000,396.225000,16.955000,25.000000
max,88.976200,100.000000,27.740000,1.000000,0.871000,8.780000,100.000000,12.126500,24.000000,711.000000,22.000000,396.900000,37.970000,50.000000


## Observations from df.describe()

| Feature | Observation |
|---------|-------------|
| `crim` | Mean ≈ 3.6 but max ≈ 89 → heavily right-skewed, extreme outliers |
| `rm` | Mean ≈ 6.28 rooms → most houses are mid-sized |
| `medv` | Mean ≈ $22.5k, Max = $50k → some houses are capped at 50 (suspicious!) |
| `age` | Mean ≈ 68.6 → most housing stock is old (built pre-1940) |
| `lstat` | Ranges 1.7% to 38% → wide inequality across neighborhoods |

**STEP 3: DATA CLEANING**



* We clean data before we analyze
* Wrong datatypes cause errors in ml models
* outliers can distort averages and model weights
* duplicates inflate your datasets and bias results


In [15]:
# Handling missing value

# isnull-- returns T/F for each cell
missing = df.isnull().sum()

# .sum ----counts missing per column
# /len(df)---- converts to %
missing_pct = (df.isnull().sum() / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Count' : missing,
    'Missing %' : missing_pct.round(2)
})

print(missing_df)
print(f"\n Total missing values in dataset: {df.isnull().sum().sum()}")

         Missing Count  Missing %
crim                 0   0.000000
zn                   0   0.000000
indus                0   0.000000
chas                 0   0.000000
nox                  0   0.000000
rm                   0   0.000000
age                  0   0.000000
dis                  0   0.000000
rad                  0   0.000000
tax                  0   0.000000
ptratio              0   0.000000
b                    0   0.000000
lstat                0   0.000000
medv                 0   0.000000

 Total missing values in dataset: 0


In [16]:
#Duplicates --- same house counted twice

print("=== Duplicate Row Check===\n")


#df.duplicated-- checks all rows in a dataframe
# return true--- row duplicate
# returns false-- if not
#.sum()---counts total row values
duplicates = df.duplicated().sum()

#prints number of row
print(f"Number of Duplicates rows: {duplicates}")


#if exists
if duplicates>0:
  print("\nDuplicate rows:")

#keep=False ---- marks all the duplicate copies as true
# r1- original
# r2-Duplicate -----  both rows will be shown

#without keep=false--only repeated copy gets true

  display(df[df.duplicated(keep=False)])

else: #no duplicates exist
  print("No duplicates Found!")

=== Duplicate Row Check===

Number of Duplicates rows: 0
No duplicates Found!


In [17]:
#dataType verification

print("===Data Type Check===\n")

print(df.dtypes)

===Data Type Check===

crim       float64
zn         float64
indus      float64
chas         int64
nox        float64
rm         float64
age        float64
dis        float64
rad          int64
tax          int64
ptratio    float64
b          float64
lstat      float64
medv       float64
dtype: object
